radixdlt/src/integration/java/com/radixdlt/consensus/simulation/BFTSimulatedTest.java

```diff
@@ -101,66 +114,95 @@ public Builder randomLatency(int minLatency, int maxLatency) {
+			this.checks.put(checkName, new LivenessCheck(duration, timeUnit));
 			return this;
 		}
 
-		public Builder checkSafety() {
-			this.checks.add(new SafetyCheck());
+		public Builder checkSafety(String checkName) {
+			this.checks.put(checkName, new SafetyCheck());
 			return this;
 		}
 
-		public Builder checkNoTimeouts() {
-			this.checks.add(new NoTimeoutCheck());
+		public Builder checkNoTimeouts(String checkName) {
+			this.checks.put(checkName, new NoTimeoutCheck());
 			return this;
 		}
 
-		public Builder checkNoSyncExceptions() {
-			this.checks.add(new NoSyncExceptionCheck());
+		public Builder checkNoSyncExceptions(String checkName) {
+			this.checks.put(checkName, new NoSyncExceptionCheck());
 			return this;
 		}
 
-		public Builder checkAllProposalsHaveDirectParents() {
-			this.checks.add(new AllProposalsHaveDirectParentsCheck());
+		public Builder checkAllProposalsHaveDirectParents(String checkName) {
+			this.checks.put(checkName, new AllProposalsHaveDirectParentsCheck());
 			return this;
 
 		}
 
 		public BFTSimulatedTest build() {
-			return new BFTSimulatedTest(ImmutableList.copyOf(nodes), latencyProvider.copyOf(), pacemakerTimeout, ImmutableList.copyOf(checks));
+			return new BFTSimulatedTest(
+				ImmutableList.copyOf(nodes),
+				latencyProvider.copyOf(),
+				pacemakerTimeout,
+				getVerticesRPCEnabled,
+				ImmutableMap.copyOf(checks)
+			);
 		}
 	}
 
 	public static Builder builder() {
 		return new Builder();
 	}
 
-	// TODO: return list of results
-	public void run(long duration, TimeUnit timeUnit) {
+	/**
+	 * Runs the test for a given time. Returns either once the duration has passed or if a check has failed.
+	 * Returns a map from the check name to the result.
+	 *
+	 * @param duration duration to run test for
+	 * @param timeUnit time unit of duration
+	 * @return map of check results
+	 */
+	public Map<String, Boolean> run(long duration, TimeUnit timeUnit) {
 		TestEventCoordinatorNetwork network = TestEventCoordinatorNetwork.builder()
 			.latencyProvider(this.latencyProvider)
 			.build();
-		SimulatedBFTNetwork bftNetwork =  new SimulatedBFTNetwork(nodes, network, pacemakerTimeout);
-		List<Completable> assertions = this.checks.stream().map(c -> c.check(bftNetwork)).collect(Collectors.toList());
-		bftNetwork.start()
-			.andThen(Completable.merge(assertions))
-			.blockingAwait(duration, timeUnit);
-		bftNetwork.stop();
+		SimulatedBFTNetwork bftNetwork =  new SimulatedBFTNetwork(nodes, network, pacemakerTimeout, getVerticesRPCEnabled);
+		List<Pair<String, Observable<Pair<String, BFTCheckError>>>> assertions = this.checks.keySet().stream()
+			.map(name -> {
+				BFTCheck check = this.checks.get(name);
+				return
+					Pair.of(
+						name,
+						check.check(bftNetwork).map(e -> Pair.of(name, e)).publish().autoConnect(2)
+					);
+			})
+			.collect(Collectors.toList());
+
+		Single<String> doneSignal = Observable.merge(assertions.stream().map(Pair::getSecond).collect(Collectors.toList()))
```

Comment:
The idea is to have the test run either until timeout or until one of the tests fail an assertion at which point we collect all current results of each test.

`done` should probably be changed to `firstError`. The name is used so that we don't stop the error check from emitting it's error to the second stream below.